### ISSUE TITLE
[FR] Add model explainability with SHAP

### ISSUE THREAD

**Issue by smurching** [mlflow/mlflow#3392]
## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [x] No. I cannot contribute this feature at this time. <- happy to help with design/code review

## Proposal Summary

[SHAP](https://github.com/slundberg/shap) is a popular library for model explanability. These explanations are useful for both:
* Understanding feature importance during model training (e.g. to guide further improvements to the model)
* Understanding the predictions made by the model on fresh data (e.g. for providing post-hoc justification for/insight into a prediction)

We should consider an extension to the MLflow APIs that simplifies logging model explanations with SHAP. 

## Motivation
See above

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### Initial Investigation [WIP]
SHAP provides an API for generating a "model explainer" given a fitted model, which can then be applied to an input dataset (`explainer.shap_values(input_dataset)`) to explain the model's predictions. For each data point, the "shap value" of each of its features is a coefficient illustrating the impact of that feature value in increasing or decreasing the model output.

For certain model types, e.g. trees ([link](https://github.com/slundberg/shap#tree-ensemble-example-with-treeexplainer-xgboostlightgbmcatboostscikit-learnpyspark-models)), SHAP is able to compute explanations using only the fitted model and model input:

```
model = xgboost.train({"learning_rate": 0.01}, xgboost.DMatrix(X, label=y), 100)

# explain the model's predictions using SHAP
# (same syntax works for LightGBM, CatBoost, scikit-learn and spark models)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)

# visualize the first prediction's explanation (use matplotlib=True to avoid Javascript)
shap.force_plot(explainer.expected_value, shap_values[0,:], X.iloc[0,:])
```

In other cases, e.g. deep learning, more robust explanations can be generated by passing "background" data constructing a model explainer ([link](https://github.com/slundberg/shap#deep-learning-example-with-deepexplainer-tensorflowkeras-models)):

```
# select a set of background examples to take an expectation over
background = x_train[np.random.choice(x_train.shape[0], 100, replace=False)]

# explain predictions of the model on four images
e = shap.DeepExplainer(model, background)
shap_values = e.shap_values(x_test[1:5])

# plot the feature attributions
shap.image_plot(shap_values, -x_test[1:5])
```

The same holds for SHAP's generic [KernelExplainer](https://github.com/slundberg/shap#model-agnostic-example-with-kernelexplainer-explains-any-function). Note that in the example below we pass a function to represent the model (``svm.predict_proba``):

```
# use Kernel SHAP to explain test set predictions
explainer = shap.KernelExplainer(svm.predict_proba, X_train, link="logit")
shap_values = explainer.shap_values(X_test, nsamples=100)

# plot the SHAP values for the Setosa output of the first instance
shap.force_plot(explainer.expected_value[0], shap_values[0][0,:], X_test.iloc[0,:], link="logit")
```

In general, it seems that constructing an explainer always requires the fitted model (or some attribute thereof) and often also an input dataset, while computing and plotting explanations requires the explainer and an input dataset. Thus in the future, we could autogenerate an explainer at model fit time (e.g. in our autologging integrations) and persist it alongside the model, so that users can later load it back and compute explanations on fresh data.

As a stepping stone, we could provide  ``mlflow.shap.log_explainer(explainer, path)`` and ``mlflow.shap.load_explainer(path)`` APIs that persist a user-constructed ``explainer`` to a specified artifact subpath and load it back for future use. Note that SHAP doesn't appear to have any built-in persistence APIs (TODO investigate, but [see docs](https://shap.readthedocs.io/en/latest/#plots)), so we may have to e.g. just pickle the explainer

**Comment by harupy** [comment]
@smurching At my previous company, I was logging shap plot and explainer like this:

```python
from contextlib import contextmanager
import os
import tempfile

import pickle
import xgboost
import shap
import matplotlib.pyplot as plt

import mlflow

# Based on:
# https://github.com/slundberg/shap#tree-ensemble-example-with-treeexplainer-xgboostlightgbmcatboostscikit-learnpyspark-models

# load JS visualization code to notebook
# shap.initjs()

# train XGBoost model
X, y = shap.datasets.boston()
model = xgboost.train({"learning_rate": 0.01}, xgboost.DMatrix(X, label=y), 100)

# explain the model's predictions using SHAP
# (same syntax works for LightGBM, CatBoost, scikit-learn and spark models)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)


# visualize the first prediction's explanation (use matplotlib=True to avoid Javascript)
plot = shap.force_plot(explainer.expected_value, shap_values[0, :], X.iloc[0, :], show=False,)
fig = shap.force_plot(
    explainer.expected_value, shap_values[0, :], X.iloc[0, :], matplotlib=True, show=False,
)

################################################################################
# Logging with MLflow
################################################################################


@contextmanager
def log_artifact_contextmanager(filename):
    with tempfile.TemporaryDirectory() as tmp_dir:
        tmp_path = os.path.join(tmp_dir, filename)
        yield tmp_path
        mlflow.log_artifact(tmp_path)


def log_shap_plot(plot, filename):
    # Note that shap plot is NOT interactive.
    assert filename.endswith(".html")
    with log_artifact_contextmanager(filename) as tmp_path:
        shap.save_html(tmp_path, plot, full_html=True)


def log_figure(fig, filename, close=True):
    with log_artifact_contextmanager(filename) as tmp_path:
        fig.savefig(tmp_path)
        if close:
            plt.close(fig)


def log_pickle(obj, filename):
    with log_artifact_contextmanager(filename) as tmp_path:
        with open(tmp_path, "wb") as f:
            pickle.dump(obj, f)


log_shap_plot(plot, "force_plot.html")
log_figure(fig, "force_plot.png")
log_pickle(explainer, "explainer.pkl")

```

**Comment by harupy** [comment]
I feel like `mlflow.shap.log_explainer(explainer, path)` will be equivalent to `log_pickle` in my code.

**Comment by harupy** [comment]
Found a discussion about how to save an explainer: https://github.com/slundberg/shap/issues/295

**Comment by eedeleon** [comment]
We have been looking into this space for a bit in our interpret-community package. Awesome to see it pop up as a top level issue!

https://github.com/interpretml/interpret-community/blob/master/python/interpret_community/mlflow/__init__.py

We targeted explanations, but explainers is another angle. I had looked at interpretable glassbox models as another approach https://github.com/interpretml/interpret/pull/123

**Comment by smurching** [comment]
To follow up on the discussion in https://github.com/slundberg/shap/issues/295 & update this thread, here's the latest from @harupy's design investigation. Namely, we've agreed on the following goals:

1. Provide an easy-to-use API (e.g. `mlflow.shap.log_explanation(model_or_predict_func, features, explanation_path=None)`) for generating and persisting model explanation plots & shap values as part of a model-training run, so that users can understand how their input features influence model predictions & behavior. Users should be able to directly pass models trained in common ML frameworks like sklearn/xgboost/keras etc, with an escape hatch of being able to pass a model-agnostic `predict_func` as the first argument to `log_explanation` if using another ML framework.

2. If possible, persist the metadata needed to explain model output on new inputs in the future. For example, after deploying a model logged with MLflow for real-time inference using e.g. MLflow's `mlflow models serve` deployment CLI, it'd be great if users could also see explanations of their model's predictions.

For 1), based on @slundberg 's helpful pointer in https://github.com/slundberg/shap/issues/295, the `shap.Explainer` API seems perfect for making use case 1) simple. One challenge we've run into is understanding which plots can & should be generated by default for an arbitrary user-supplied model, using the `shap.Explanation` objects returned by `shap.Explainer`; the easiest seems to be generating summary bar plots (`shap.bar_plot`). It seems with additional input from the user (e.g. which label to visualize for multiclass classification), we could additionally generate `shap.summary_plot`.


For 2), our current understanding is that although it's possible to generate explanations solely using a model and new serving input, e.g:
```
import shap
explainer = shap.Explainer(deployed_model, single_serving_input_example)
shap_values = explainer(single_serving_input_example)
# generate plots etc
```

the discussion in https://github.com/slundberg/shap/issues/370 and https://github.com/slundberg/shap/issues/375 suggests the quality of these explanations may not be great (i.e. supplying more examples when constructing an `Explainer` results in higher quality explanations). One solution could be logging the input `features` to `log_explanation` along with the generated explanation plots & shap values.

**Comment by smurching** [comment]
@slundberg and others, we'd love your feedback/thoughts on:

1) What's the long-term goal for `shap.Explainer` and `shap.Explanation`? The former seems aimed at abstracting away reasoning about the model type (is it a tree model, deep learning model, etc) from the user. Is the goal of `shap.Explanation` to provide a similar model-agnostic interface that can be consumed by SHAP's plotting APIs and other tools?
2) Is our understanding correct that several (is there a good way to reason about how many?) background examples are needed in order to generate accurate explanations on new inputs at model-serving time?

And @eedeleon sorry I missed your earlier comment - I think @harupy independently stumbled upon InterpretML, found that it had an API very similar to SHAP, and that it generated very useful visualizations :)

If you & @slundberg are up for it, it'd be great to chat some time to see how we can collaborate - thanks again to both of you for your thoughts/input on this!

**Comment by ademarslan** [comment]
@harupy is there any work that you implemented the summary plot to create artifacts and log to the mlflow?
